src/lxc/lxccontainer.c

```diff
@@ -4659,6 +4680,131 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
 
 WRAP_API_2(bool, lxcapi_restore, char *, bool)
 
+static int do_lxcapi_mount(struct lxc_container *c,
+			const char *source, const char *target,
+			const char *filesystemtype, unsigned long mountflags,
+			const void *data, struct lxc_mount *mnt) {
+	int ret;
+	int len;
+	int fd;
+	int status;
+	char* pref = NULL;
+	char* temp;
+	char *suff;
+	char *path;
+	pid_t pid;
+	pid_t init_pid;
+	struct stat sb;
+
+	// Create a name template
+	// Do it here to be able to use temp in both parent and child
+	pref = getenv("LXC_DIR");
+	if (pref == NULL)
+		pref = "/var/lib/lxc";
+
+	len = strlen(pref) + strlen("shmounts") + strlen(c->name) + strlen("lxcmountXXXXXX") + 4;
+	temp = (char *)malloc(len);
+	ret = snprintf(temp, len, "%s/shmounts/%s/lxcmountXXXXXX", pref, c->name);
+	if (ret < 0 || ret >= len) {
+		ERROR("Error writing shmounts tempdir name");
+		return -1;
+	}
+
+	// Create a temporary dir / file
+	// **under the shared mountpoint**
+	ret = stat(source, &sb);
+	if (ret < 0)
+		return -1;
+
+	if (S_ISREG(sb.st_mode)) {
+		fd = mkstemp(temp);
+		if (fd < 0) {
+			ERROR("Error creating shmounts temporary file");
+			return -1;
+		}
+	} else if (S_ISDIR(sb.st_mode)) {
+		temp = mkdtemp(temp);
+		if (temp == NULL) {
+			ERROR("Error creating shmounts tempdir");
+			return -1;
+		}
+	} else {
+		ERROR("Unexpected type of shared mountpoint");
+		return -1;
+	}
+
+	// Do the fork
+	pid = fork();
+	if (pid < 0)
+		return -1;
+
+	if (pid == 0) {
+		// Do the mount
+		ret = mount(source, temp, filesystemtype, mountflags, data);
+		if (ret < 0)
+			return -1;
+
+		init_pid = do_lxcapi_init_pid(c);
+		if (init_pid < 0)
+			return -1;
+
+		// Enter the container namespaces
+		ret = setns(init_pid, CLONE_NEWNS);
+		if (ret < 0)
+			return -1;
+		// if fully unprivileged
+		if (am_guest_unpriv() && !lxc_list_empty(&c->lxc_conf->id_map)) {
+			ret = setns(init_pid,CLONE_NEWUSER);
```

Comment:
You need to attach to the user namespace first and then to the mount namespace. Otherwise the kernel will refuse the `setns()` to the mount namespace.